book/src/dev/rfcs/0006-contextual-difficulty.md

```diff
@@ -0,0 +1,302 @@
+  of the blocks in the averaging window.
+
+**median timespan**: The average number of seconds taken to generate the blocks
+  in the averaging window. Calculated using the difference of median block spans
+  in and after the averaging window, then damped and bounded.
+
+**target timespan**: The target spacing for an averaging window's worth of
+  blocks.
+
+Consensus:
+* **consensus rule:** A protocol rule which all nodes must apply consistently,
+                      so they can converge on the same chain fork.
+
+* **structural/semantic/contextual verification**: as defined in [RFC2].
+
+State:
+* **block chain**: A sequence of valid blocks linked by inclusion of the
+  previous block hash in the subsequent block. Chains are rooted at the
+  genesis block and extend to a tip.
+
+* **relevant chain**: The relevant chain for a block starts at the previous
+  block, and extends back to genesis.
+
+* **relevant tip**: The tip of the relevant chain.
+
+* **non-finalized state**: State data corresponding to blocks above the reorg
+  limit. This data can change in the event of a chain reorg.
+
+* **finalized state**: State data corresponding to blocks below the reorg
+  limit. This data cannot change in the event of a chain reorg.
+
+* **non-finalized tips**: The highest blocks in each non-finalized chain. These
+  tips might be at different heights.
+
+* **finalized tip**: The highest block in the finalized state. The tip of the best
+  chain is usually 100 blocks (the reorg limit) above the finalized tip. But it can
+  be lower during the initial sync, and after a chain reorganization, if the new
+  best chain is at a lower height.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The difficulty threshold for the next block is calculated using the difficulty
+thresholds and times of recent blocks. Zcash uses the most recent 28 blocks in
+the **relevant chain** in its difficulty adjustment calculations.
+
+The difficulty adjustment calculations adjust the **mean target difficulty**,
+based on the difference between the **median timespan** and the
+**target timespan**.
+
+Since contextual validation is only used for post-Sapling blocks, we can assume
+that there will be at least 28 blocks in any relevant chain.
+
+Difficulty threshold calculations are performed using unsigned 256-bit integers.
+Time calculations are performed using unsigned 32-bit integers.
+
+TODO:
+  - check how zcashd implements signed median time differences
+  - open a ticket to update the Zcash spec
+
+## The relevant chain
+[relevant-chain]: #relevant-chain
+
+The relevant chain can be retrieved from the state service [RFC5] as follows:
+* if the previous block is the finalized tip:
+  * get recent blocks from the finalized state
+* if the previous block is in the non-finalized state:
+  * get recent blocks from the relevant chain, then
+  * get recent blocks from the finalized state, if required
```

Comment:
Good point.

I'll update the RFC so that the changes to the state service are separate from the new contextual validation functions.

Reply:
The validation is done during `CommitBlock` in this RFC.

There might be some future work where we pre-download and pre-validate headers, before downloading the entire block. That would require an extra state service request, but the request would just call the existing state service function used for contextual validation.

I'll sketch that design in the future work section.

Reply:
Fixed in 02dd6bc4b7fe2ee243ffbf8d2ee774a568d07e08, turns out there are actually 2 separate contextual checks: the difficulty threshold, and the difficulty filter.

(We are checking the difficulty filter in zebra-consensus, but that's wrong, because testnet minimum difficulty blocks make it a contextual check. I'll fix that in a new PR.)